### ISSUE TITLE
"GPJWP: Ignoring module without a found package.json"

### ISSUE THREAD

**Issue by brunobely** [lostpebble/generate-package-json-webpack-plugin#45]
I'm not sure if this is the right place to ask, as I can't pinpoint what's causing the error I see. Hopefully it is.

I set up a NestJS app that I build with Webpack and deploy to Google Cloud Functions and it works fine until I try to import another module from the same workspace (`"@proto/core": "workspace:*"` in my `package.json`). This is the error log I see when trying to build without Webpack.

```sh
❯ nest build
src/metrics/metrics.service.ts:4:27 - error TS2307: Cannot find module '@proto/core' or its corresponding type declarations.

4 import { Account } from '@proto/core';
                            ~~~~~~~~~~~~~~~~~

Found 1 error(s).
```

And if I build with webpack (which uses `generate-package-json-webpack-plugin`, see config below):

```sh
❯ nest build --webpack --webpackPath webpack.config.cjs

GPJWP: Ignoring module without a found package.json: @proto/core ("@proto/core/package.json" couldn't resolve): Cannot find module '@proto/core'
Require stack:
- /Users/path/to/monorepo/node_modules/.pnpm/generate-package-json-webpack-plugin@2.6.0/node_modules/generate-package-json-webpack-plugin/index.js
- /Users/path/to/monorepo/apps/api/webpack.config.cjs
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/actions/build.action.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/actions/index.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/commands/command.loader.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/commands/index.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/bin/nest.js
GPJWP: Couldn't resolve a version for module "@proto/core" (from portable ID: external commonjs "@proto/core") - will be ignored in final output.

ERROR in /Users/path/to/monorepo/apps/api/src/metrics/metrics.service.ts
4:26-43
[tsl] ERROR in /Users/path/to/monorepo/apps/api/src/metrics/metrics.service.ts(4,27)
      TS2307: Cannot find module '@proto/core' or its corresponding type declarations.
ts-loader-default_e3b0c44298fc1c14

webpack 5.74.0 compiled with 1 error in 5172 ms
```

Which to me seems like if only Webpack could find the `@proto/core` package then it would all go away? I'm not totally sure.

Things I've tried:
- adding `sourcePackageFilenames` in the `GeneratePackageJsonPlugin` call
- adding the package in question to `additionalDependencies` in the `GeneratePackageJsonPlugin` call (even though it's deprecated)
- adding the package in question to `dependency` to the `basePackage`
- adding the package in question to `devDependency` to the `basePackage`

None of which seem to make a difference.

Below is my Webpack config:

```js
const GeneratePackageJsonPlugin = require('generate-package-json-webpack-plugin');
const path = require('path');

const basePackage = {
  name: 'api',
  version: '0.0.1',
  main: './main.js',
  devDependencies: {
    '@proto/core': 'workspace:*',
  },
  engines: {
    node: '>= 14',
  },
};

module.exports = {
  entry: './src/main.ts',
  mode: process.env.NODE_ENV,
  module: {
    rules: [
      {
        test: /\.ts$/,
        use: 'ts-loader',
        exclude: /node_modules/,
      },
    ],
  },
  output: {
    filename: 'main.js',
    path: path.resolve(__dirname, 'dist'),
  },
  plugins: [
    new GeneratePackageJsonPlugin(basePackage, {
      sourcePackageFilenames: [
        path.join(__dirname, '../../../packages/proto/core/package.json'),
      ],
    }),
  ],
  resolve: {
    extensions: ['.ts', '.js'],
  },
  target: 'node',
};
```

**Comment by lostpebble** [comment]
This error is being caused by the root issue at hand here:

```
Cannot find module '@proto/core' or its corresponding type declarations
```

You most likely haven't set up your monorepo correctly- as there should be a symlinked folder (generally) of your workspace module inside `/node_modules` - and this would allow bundlers / builders who are looking for this dependency to find it.

The same applies to this plugin- if the module cannot be found in the expected way (i.e. searching in all parent `/node_modules` folders) then we can't find the exact version being used either- since we can't find the `package.json` file to pull that version number from.

`sourcePackageFilenames` is being used incorrectly here- it is an old option, which shouldn't really be used anymore. But the way it was used was to define your own project's `package.json` files, in which we would source the versions of your dependencies from. Please ignore that option rather- the default way the plugin works is best.

Basically, you need to make sure that whatever is running your code (be it for building or for runtime)- that it can find that dependency its looking for (`@proto/core`). I use yarn and with their workspaces it seems like it does this all for me.

**Comment by brunobely** [comment]
Thanks for looking into this! This makes sense. I was able to fix this error:

```
Cannot find module '@proto/core' or its corresponding type declarations
```

The package is now in `node_modules` and Webpack runs successfully, but the package is still being ignored (so the built code fails to find it once run). Any ideas? This is the log when I run:

```
GPJWP: Ignoring module without a found package.json: @proto/core ("@proto/core/package.json" couldn't resolve): Cannot find module '@proto/core'
Require stack:
- /Users/path/to/monorepo/node_modules/.pnpm/generate-package-json-webpack-plugin@2.6.0/node_modules/generate-package-json-webpack-plugin/index.js
- /Users/path/to/monorepo/apps/api/webpack.config.cjs
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/actions/build.action.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/actions/index.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/commands/command.loader.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/commands/index.js
- /Users/path/to/monorepo/node_modules/.pnpm/@nestjs+cli@9.1.5_webpack-cli@4.10.0/node_modules/@nestjs/cli/bin/nest.js
GPJWP: Couldn't resolve a version for module "@proto/core" (from portable ID: external commonjs "@proto/core/ts/models/account.pb") - will be ignored in final output.
webpack 5.74.0 compiled successfully in 4707 ms
```

**Comment by brunobely** [comment]
Added the `@proto/core` package to the [`externals` whitelist](https://github.com/lostpebble/generate-package-json-webpack-plugin#important-note-on-externals) in the Webpack config (which makes sense, since I want this package bundled with the code) and it solved the issue!

Thank you again.
